doc/cephfs/capabilities.rst

```diff
@@ -110,3 +110,123 @@ capabilities. For example:
+---------------------------------------------------------
+In MDS there are four different locks for each inode, they are simplelock,
+scatterlock, filelock and locallock. For each lock it have several different
+lock states, and in each state it will have different meaning and will issue
+the relevant capabilities to the clients.
+
+The lock state structure:
+
+::
+
+  struct sm_state_t {
+    /*
+     * The next state it will transfer to, 0 means stable, all the others will
+     * be interim transition states just before transfering to the next stable
+     * state.
+     */
+    int next;
+
+    /*
+     * Is this lock state in loner mode ? If yes, only the loner client could
+     * be issued the capabilities.
+     */
+    bool loner;
+
+    /* What's the lock state it will be for the replica MDSes */
+    int replica_state;
+
+    char can_read;
+    char can_read_projected;
+    char can_rdlock;
+    char can_wrlock;
+    char can_force_wrlock;
+    char can_lease;
+    char can_xlock;
+
+    /*
+     * For auth MDS, this is what capabilities are allowed to be issued to
+     * all the clients in this state.
+     */
+    int caps;
+
+    /*
+     * For auth MDS, this is what capabilities are allowed to be issued only
+     * to the loner client in this state, if "loner_caps" is none zero then
+     * the above "loner" member must be true, and all the 3 other caps related
+     * members will be zero.
+     */
+    int loner_caps;
+
+    /*
+     * If one client_request will xlock the lock in MDS side, except the "caps"
+     * this is the extra capabilities are allowed to be issued only to the
+     * relevant client.
+     */
+    int xlocker_caps;
+
+    /*
+     * If current inode is in replica MDSes, this is what the only capabilities
+     * are allowed to be issued to all the clients from replica MDSes.
+     */
+    int replica_caps;
+  };
+
+In each state the MDS Locker will always try to issue all the capabilities to the
+clients allowed, even some capabilities are not needed or wanted by the clients,
+the pre-issuing capabilities could reduce latency in some cases.
+
+If there is only one client, ususally it will be the loner client for all the inodes.
+While in multiple clients case, the MDS will try to caculate a loner client out for
+each inode depending on the capabilities the clients (needed | wanted), but usually
+it will fail. The loner client will always get all the capabilities.
+
+The filelock and the capabilities
+---------------------------------
+The filelock will control files' partial metadatas' and the file contents' access
+permissions. The metadatas include **mtime**, **atime**, **size**, etc.
+
+**Fs**: Once a client has it, all the other clients are can't have to have **Fw**.
+
+**Fx**: Only the loner client could have this capability. The lock state will be transferred
+        to LOCK_EXCL finally and the loner client could have all the file capabilities.
```

Comment:
Not sure if this is accurate, but I'm trying to clean up the wording a bit:

Only the loner client is allowed this capability. Once the lock state transitions to LOCK_EXCL, the loner client is granted this along with all other file capabilities.

Reply:
Done and fixed them all, thanks @jtlayton 

For this with a minor fix:

`s/all other file capabilities/all other file capabilities except the **Fl** capability/`